File: src/dataset/c/bob_Exercism_m3g4d1v3r/gemini/GeminiFlash_bob_v4.c
#include "bob.h"
#include <ctype.h>
#include <stdbool.h>

bool ends_with_question_mark(char *chr) {
    if (*chr == '\0') {
        return false;
    }

    char *end = chr;
    while (*end != '\0') {
        end++;
    }
    end--;

    while (end >= chr && isspace((unsigned char)*end)) {
        end--;
    }

    if (end < chr) {
        return false;
    }

    return (*end == '?');
}

bool has_all_capital_letters(char *chr) {
    bool capital_found = false;
    while (*chr != '\0') {
        if (islower((unsigned char)*chr)) {
            return false;
        }
        if (isupper((unsigned char)*chr)) {
            capital_found = true;
        }
        chr++;
    }
    return capital_found;
}

bool silent_message(char *chr) {
    while (*chr != '\0') {
        if (isspace((unsigned char)*chr) == 0) {
            return false;
        }
        chr++;
    }
    return true;
}

char *hey_bob(char *greeting) {
    bool question = ends_with_question_mark(greeting);
    bool capital = has_all_capital_letters(greeting);
    bool silent = silent_message(greeting);

    if (question && !capital) {
        return "Sure.";
    } else if (!question && capital) {
        return "Whoa, chill out!";
    } else if (question && capital) {
        return "Calm down, I know what I'm doing!";
    } else if (silent) {
        return "Fine. Be that way!";
    } else {
        return "Whatever.";
    }
}

File: src/dataset/javascript/bob_exercism-javascript-ffflorian/gemini/GeminiFlash_bob_v4.js
export function hey(input) {
  // Trim the input once at the beginning to remove leading/trailing whitespace.
  // This operation is necessary for correct functionality and creates a new string.
  const trimmedInput = input.trim();

  // Handle the empty string case early.
  // This is a direct check and is very efficient.
  if (trimmedInput.length === 0) {
    return 'Fine. Be that way!';
  }

  // Pre-compute the uppercase and lowercase versions of the trimmed input.
  // This avoids redundant string transformations later, reducing CPU cycles and memory allocations
  // that would otherwise occur from multiple calls to .toUpperCase() and .toLowerCase().
  const trimmedInputUpperCase = trimmedInput.toUpperCase();
  const trimmedInputLowerCase = trimmedInput.toLowerCase();

  // Determine if the input is "shouting".
  // A string is considered shouting if all its alphabetic characters are uppercase
  // AND it contains at least one alphabetic character (checked by comparing upper and lower case).
  // This approach is robust and efficient.
  const isShouting = (trimmedInput === trimmedInputUpperCase) && (trimmedInputUpperCase !== trimmedInputLowerCase);

  // Determine if the input ends with a question mark.
  // Accessing the character directly by index (`[length - 1]`) is generally more
  // performant than `slice(-1)` or `endsWith('?')` for single-character checks,
  // as it avoids creating a new temporary string object.
  // Since `trimmedInput.length === 0` is already handled, we can safely access `trimmedInput.length - 1`.
  const endsWithQuestionMark = trimmedInput[trimmedInput.length - 1] === '?';

  // Process the input based on the determined characteristics.
  if (isShouting) {
    if (endsWithQuestionMark) {
      return "Calm down, I know what I'm doing!";
    }
    return 'Whoa, chill out!';
  }

  if (endsWithQuestionMark) {
    return 'Sure.';
  }

  // Default response if none of the above conditions are met.
  return 'Whatever.';
}